# Shell

Unix shell — командный интерпретатор, используемый в операционных системах семейства Unix, в котором пользователь может либо давать команды операционной системе по отдельности, либо запускать скрипты, состоящие из списка команд

![shell](https://habrastorage.org/files/158/0a1/e66/1580a1e669284c40917ae98946490652.png)

#### Подробнее в статье: https://habr.com/post/267825/

In [7]:
!hostname

vmazaev-nix


In [8]:
!whoami

vmazaev


In [9]:
!pwd

/home/vmazaev/shad


In [10]:
!ls /bin

bash	       getfacl	   nisdomainname     su
bunzip2        grep	   ntfs-3g	     sync
busybox        gunzip	   ntfs-3g.probe     systemctl
bzcat	       gzexe	   ntfs-3g.secaudit  systemd
bzcmp	       gzip	   ntfs-3g.usermap   systemd-ask-password
bzdiff	       hciconfig   ntfscat	     systemd-escape
bzegrep        hostname    ntfscluster	     systemd-hwdb
bzexe	       ip	   ntfscmp	     systemd-inhibit
bzfgrep        journalctl  ntfsfallocate     systemd-machine-id-setup
bzgrep	       kbd_mode    ntfsfix	     systemd-notify
bzip2	       keyctl	   ntfsinfo	     systemd-tmpfiles
bzip2recover   kill	   ntfsls	     systemd-tty-ask-password-agent
bzless	       kmod	   ntfsmove	     tailf
bzmore	       less	   ntfstruncate      tar
cat	       lessecho    ntfswipe	     tempfile
chacl	       lessfile    open		     touch
chgrp	       lesskey	   openvt	     true
chmod	       lesspipe    pidof	     udevadm
chown	       ln	   ping		     ulockmgr_server
chvt	       loadkeys    ping6	     umount
cp	 

#### history

In [1]:
!man

What manual page do you want?


In [ ]:
!info

#### permissions

In [13]:
!chmod
!chown

chmod: missing operand
Try 'chmod --help' for more information.
chown: missing operand
Try 'chown --help' for more information.


#### stdin, stdout, pipes

In [16]:
!date

Sat Nov 17 18:42:12 MSK 2018


In [15]:
!date > tmp

In [18]:
!cat tmp | wc

      1       6      29


In [19]:
!date >> tmp

In [20]:
!cat tmp | wc

      2      12      58


In [22]:
!wc < tmp

 2 12 58


# SSH

![ssh](https://www.ssh.com/s/ssh-tunneling-3138x956-4zmvrU8b.png)

#### Для начала: https://guides.hexlet.io/ssh/
#### Для продвинутых: https://habr.com/post/122445/ & https://xakep.ru/2016/12/16/ssh-tips-n-tricks/ (по подписке)
#### ssh-agent: http://xgu.ru/wiki/Управление_ключами_SSH_с_помощью_агента (в ubuntu входит в состав gnome-keyring-daemon, запущен по-умолчанию)

In [23]:
!ls -l ~/.ssh

total 440
-rw-rw-r-- 1 vmazaev vmazaev    486 Oct 27 21:53 config
-rw------- 1 vmazaev vmazaev   1675 Apr 16  2018 docean_rsa
-rw-r--r-- 1 vmazaev vmazaev    401 Apr 16  2018 docean_rsa.pub
-rw------- 1 vmazaev vmazaev   1679 Apr  3  2018 id_rsa
-rw-r--r-- 1 vmazaev vmazaev    401 Apr  3  2018 id_rsa.pub
-rw------- 1 vmazaev vmazaev 224630 Oct  4 14:57 known_hosts
-rw------- 1 vmazaev vmazaev 198744 Apr  6  2018 known_hosts.old


In [30]:
!tail -6 ~/.ssh/config | sed -r 's/([0-9]+)/XXX/g'

Host docean
	HostName XXX.XXX.XXX.XXX
	User root
	ForwardAgent yes
	IdentityFile ~/.ssh/docean_rsa



In [ ]:
!ssh 11.22.33.44

In [ ]:
!ssh -i ~/key.pem ubuntu@11.22.33.44

In [ ]:
!man ssh
!man ssh_config

# curl

![curl](https://deividsdocs.files.wordpress.com/2018/05/captura-e1527287262746.png?w=768)

#### Почитать: https://losst.ru/kak-polzovatsya-curl

In [34]:
!curl "yandex.ru" -o body

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0


In [1]:
!curl "https://api.telegram.org/bot${BOT_TOKEN}/getMe"

{"ok":true,"result":{"id":648896946,"is_bot":true,"first_name":"test","username":"my_super_tetris_bot"}}

In [11]:
!curl -X POST "https://api.telegram.org/bot${BOT_TOKEN}/sendPhoto" -H "Content-Type: application/json" \
    -d '{"chat_id": "122377527", "photo": "https://bit.ly/2OPlbCl"}'

{"ok":true,"result":{"message_id":5,"from":{"id":648896946,"is_bot":true,"first_name":"test","username":"my_super_tetris_bot"},"chat":{"id":122377527,"first_name":"Vadim","last_name":"Mazaev","username":"vadim_mazaev","type":"private"},"date":1542735127,"photo":[{"file_id":"AgADBAADc6gxG3J-hVMa6lTuGJSQPztHoBoABKfaaYVBqpiJbtsEAAEC","file_size":1480,"width":90,"height":51},{"file_id":"AgADBAADc6gxG3J-hVMa6lTuGJSQPztHoBoABL7IRi7LETiLb9sEAAEC","file_size":20283,"width":320,"height":180},{"file_id":"AgADBAADc6gxG3J-hVMa6lTuGJSQPztHoBoABHlqHuoN7X_CcNsEAAEC","file_size":81617,"width":800,"height":450},{"file_id":"AgADBAADc6gxG3J-hVMa6lTuGJSQPztHoBoABO2RjJ8Qin47cdsEAAEC","file_size":171423,"width":1280,"height":720},{"file_id":"AgADBAADc6gxG3J-hVMa6lTuGJSQPztHoBoABNbFTAABK5mPiXLbBAABAg","file_size":303690,"width":1920,"height":1080}]}}

In [12]:
!curl -X POST "https://api.telegram.org/bot${BOT_TOKEN}/sendPhoto" -H "Content-Type: multipart/form-data" \
    -F 'chat_id=122377527' -F 'photo=@/home/vmazaev/kiminonawa_2.jpg'

{"ok":true,"result":{"message_id":6,"from":{"id":648896946,"is_bot":true,"first_name":"test","username":"my_super_tetris_bot"},"chat":{"id":122377527,"first_name":"Vadim","last_name":"Mazaev","username":"vadim_mazaev","type":"private"},"date":1542735359,"photo":[{"file_id":"AgADAgADx6kxG_TJoUtSJBOUmcoV3OQ_8w4ABIPeOUYBBzLKwzUDAAEC","file_size":1349,"width":90,"height":42},{"file_id":"AgADAgADx6kxG_TJoUtSJBOUmcoV3OQ_8w4ABJ78BUv90N5vxDUDAAEC","file_size":22620,"width":320,"height":150},{"file_id":"AgADAgADx6kxG_TJoUtSJBOUmcoV3OQ_8w4ABMeh3txD589FxTUDAAEC","file_size":117510,"width":800,"height":375},{"file_id":"AgADAgADx6kxG_TJoUtSJBOUmcoV3OQ_8w4ABLSQW72TgjJzxjUDAAEC","file_size":280797,"width":1280,"height":600},{"file_id":"AgADAgADx6kxG_TJoUtSJBOUmcoV3OQ_8w4ABI-R57ftZ7S5xzUDAAEC","file_size":714984,"width":2000,"height":938}]}}

#### multipart/form-data: https://ru.wikipedia.org/wiki/Multipart/form-data

In [7]:
!wget "https://www.python.org/ftp/python/3.7.1/Python-3.7.1.tar.xz"

--2018-11-17 22:24:32--  https://www.python.org/ftp/python/3.7.1/Python-3.7.1.tar.xz
Resolving www.python.org (www.python.org)... 2a04:4e42:9::223, 151.101.36.223
Connecting to www.python.org (www.python.org)|2a04:4e42:9::223|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 16960060 (16M) [application/octet-stream]
Saving to: ‘Python-3.7.1.tar.xz’

Python-3.7.1.tar.xz 100%[===================>]  16.17M  8.42MB/s    in 1.9s    

2018-11-17 22:24:34 (8.42 MB/s) - ‘Python-3.7.1.tar.xz’ saved [16960060/16960060]



# tmux

![tmux](https://cdn-images-1.medium.com/max/1200/1*jzpJVBT31BFW9ykUBDG7cQ.png)

#### Шпаргалка по tmux: https://habr.com/post/126996/

## If you just want to keep bot running when you leave ssh session

In [ ]:
!nohup python3 main.py &

# Network things

## DNS

![dns](http://geek-nose.com/wp-content/uploads/2018/05/what-is-dns.png)

In [9]:
!host ya.ru

ya.ru has address 87.250.250.242
ya.ru has IPv6 address 2a02:6b8::2:242
ya.ru mail is handled by 10 mx.yandex.ru.


## Common TCP ports (listening)

1. HTTP - 80
2. HTTPS - 443
3. SSH - 22

## What happens when you type an url in the browser and press enter?
#### Лайт-вариант: https://medium.com/@maneesha.wijesinghe1/what-happens-when-you-type-an-url-in-the-browser-and-press-enter-bb0aa2449c1a
#### Ultra hard mode: https://github.com/alex/what-happens-when

# How to run your bot on remote server

0. AWS: see telegram channel for instruction
1. Submit your code to gitlab
2. Ssh to your server
3. Clone your git repo, i.e. `git clone git@gitlab.manytask ...`
4. Change branch to `submits/cinemabot`, i.e. `git checkout -b `
5. Install requirements, i.e. `pip install -r requirements.txt`
6. Run tmux
7. Start your bot, i.e. `python3 main.py`
8. Detach from tmux

### Update code
1. Submit your code to gitlab
2. Ssh yo yur server
3. Pull changes from gitlab, i.e. `git pull origin submits/cinemabot`
4. Install new requirements if needed
5. Attach to tmux, restart bot, detach from tmux

# Docker (basic)

![docker](https://vidikon.info/wp-content/uploads/2018/01/docker-banner-1920x500.png)

#### Документация: https://docs.docker.com/